src/main/GCNodeAllocator.cpp

```diff
@@ -0,0 +1,499 @@
+/*
+ *  R : A Computer Language for Statistical Data Analysis
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/AddressSanitizer.hpp"
+#include "rho/AllocationTable.hpp"
+#include "rho/GCNodeAllocator.hpp"
+
+#ifdef HAVE_ADDRESS_SANITIZER
+// Quarantine free lists are used to store freed objects for a while before
+// they can be reused. Allocations are poisoned while in the quarantine. This
+// makes it more likely that Address Sanitizer can detect use-after-free
+// errors.
+rho::FreeListNode* rho::GCNodeAllocator::s_quarantine[s_num_freelists];
+#endif
+
+// The allocation table tracks medium and large allocations.
+rho::AllocationTable* rho::GCNodeAllocator::s_alloctable = nullptr;
+
+// Superblock pools by size class. Each superblock in linked in this
+// array has at least one untouched block available.
+rho::AllocatorSuperblock* rho::GCNodeAllocator::s_superblocks[
+    s_num_small_pools + s_num_medium_pools];
+
+// Tracking heap bounds for fast rejection in pointer lookup.
+uintptr_t rho::GCNodeAllocator::s_heap_start = UINTPTR_MAX;
+uintptr_t rho::GCNodeAllocator::s_heap_end = 0;
+
+// Free lists head pointers.
+rho::FreeListNode* rho::GCNodeAllocator::s_freelists[s_num_freelists];
+
+#ifdef ALLOCATION_CHECK
+// Helper function for allocator consistency checking.
+// An additional allocation map is added which shadows the state of the
+// allocator. The extra allocation map is checked to verify each operation
+// on the allocator.
+
+static void add_to_allocation_map(void* allocation, std::size_t size);
```

Comment:
Nit: The C++ idiom to get file-scoped declarations is to enclose them in anonymous namespace, instead of declaring them static.